flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -381,13 +386,15 @@ void onConsumedSubpartition(int subpartitionIndex) {
 			return;
 		}
 
-		int refCnt = pendingReferences.decrementAndGet();
+		// externally managed partitions are released on-demand and do not use ref-counting
+		if (!isManagedExternally) {
+			int refCnt = pendingReferences.decrementAndGet();
 
-		if (refCnt == 0) {
-			partitionManager.onConsumedPartition(this);
-		}
-		else if (refCnt < 0) {
-			throw new IllegalStateException("All references released.");
+			if (refCnt == 0) {
+				partitionManager.onConsumedPartition(this);
+			} else if (refCnt < 0) {
+				throw new IllegalStateException("All references released.");
+			}
```

Comment:
Looking at the changes to the `ResultPartition` I'm was wondering whether it wouldn't make sense to have a `ReleaseOnConsumptionResultPartition` implementation which extends `ResultPartition`. Now we are mixing the logic for the implicit and explicit release in one class.

Reply:
I wouldn't be against introducing an additional ResultPartition class. It is fairly straight-forward to do, and outright prevents anyone from interfering with the refCounts if it is externally managed, which is quite useful.